src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeWriter.cs

```diff
@@ -135,15 +135,16 @@ private void AllocateMemory(int sizeHint)
                         _tailBytesBuffered = 0;
                     }
 
-                    BufferSegment newSegment = AllocateSegment(sizeHint);
+                    int newSegmentSize = Math.Min(PipeOptions.MaximumSegmentSize, _tail.Capacity * 2);
```

Comment:
This is simple to understand but another possible technique could be to base the growth on number of segments rather than the last segment (so taking the speed of the consumer into account to shrink the next segment). I'm not sure if it's a big concern though because:
- If the last segment is ever fully consumed, it'll restart to 0.
- If the last segment isn't fully consumed because the reader can't keep up, the pause threshold can be adjusted and working with bigger blocks of memory is better for the reader.